### ISSUE TITLE
Evaluated command doesn't recognize assignment

### ISSUE THREAD

**Issue by beltaurus** [koalaman/shellcheck#1777]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2154
- My shellcheck version (`shellcheck --version` or 'online'): online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related
https://github.com/koalaman/shellcheck/issues/1758 does report a similar issue, though

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
eval a=DEFINED
b=$a
echo "$b"
```

#### Here's what shellcheck currently says:
```
Line 3	SC2154: a is referenced but not assigned.
```
#### Here's what I wanted or expected to see:

Nothing (recognize that 'a' was assigned in eval'ed command...)

**Comment by Gandalf-** [comment]
This seems closely related to https://github.com/koalaman/shellcheck/issues/1724, the same kind of "command aliasing". 
Maybe we can fix these at the same time.

**Comment by koalaman** [comment]
You're obviously right, but this is intended behavior. Here's' the rationale from the semi-related  [SC2034](https://github.com/koalaman/shellcheck/wiki/SC2034) wiki page: 

---
#### Indirection

It's ShellCheck's intended behavior to emit this warning for any variable that is only referenced though indirection:

    # foo generates a warning, even though it has five indirect references
    foo=42     
    name=foo
    echo "${!name} $((name))"
    export "$name"; eval "echo $name"
    declare -n name; echo "$name"

This is an intentional design decision and not a bug. If you have variables that will not have direct references, consider using an associative array in bash, or just [[Ignore]] the warning. 

Tracking indirect references is a common problem for compilers and static analysis tool, and it is known to be unsolvable in the most general case. There are two ways to handle unresolved indirections (which in a realistic program will essentially be all of them):

* Avoid false positives by assuming all variables may potentially be used (disable all unused variable warnings)
* Keep true positives by allowing some false positives (emit unused variable warnings that may be incorrect)

Compilers are forced to do the former, but static analysis tools generally do the latter. This includes `pylint`, `jshint` and `shellcheck` itself. This is a design decision meant to make the tools more helpful at the expense of some noise. For consistency and to avoid giving the impression that it should work more generally, ShellCheck does not attempt to resolve even trivial indirections.
